jsonmunge.1.md
%jsonmunge(1) user manual | version 1.3.5 f86e208
% R. S. Doiel
% 2026-02-12

# NAME

jsonmunge 

# SYNOPSIS

jsonmunge [OPTIONS] TEMPLATE_FILENAME

# DESCRIPTION

jsonmunge is a command line tool that takes a JSON document and
one or more Go templates rendering the results. Useful for
reshaping a JSON document, transforming into a new format,
or filter for specific content.

- TEMPLATE_FILENAME is the name of a Go text tempate file used to render
  the outbound JSON document

# OPTIONS

-help
: display help

-license
: display license

-version
: display version

-E, -expression
: use template expression as template

-i, -input
: input filename

-nl, -newline
: if true add a trailing newline

-o, -output
: output filename

-quiet
: suppress error messages


# EXAMPLES

If person.json contained

~~~
   {"name": "Doe, Jane", "email":"jd@example.org", "age": 42}
~~~

and the template, name.tmpl, contained

~~~
   {{- .name -}}
~~~

Getting just the name could be done with

~~~
    cat person.json | jsonmunge name.tmpl
~~~

This would yield

~~~
    "Doe, Jane"
~~~

jsonmunge 1.3.5



RELEASES.md

Release Process and tags
========================

Preleases and production releases
---------------------------------

This describes a simple release process organized around semantic versioned tags.

When a new release is ready it should be tag as v0.XX.XX-pre (where XX is a number) and published as a 'pre-release' on
Github. Send a note to the development group so someone other than the proposer of the release will be able to 
indepentently evaluate the release changes.  At a minimum they should run `bash test_cmds.bash` and walk through some 
of the documentation examples. If the release is verified as ready then a new release will be cut with a tag like
v0.XX.XX (there the XX are the same as in v0.XX.XX-pre). E.g. v0.0.10-pre would become v0.0.10 on success.

If the release fails verification, bugs need to be fixed and a new release proposed create after the fixes. 
The "patch" number persion should be incremented in v0.XX.XX-pre (e.g. v0.0.10-pre would be followed by 
v0.0.11-pre indicating that patches). 

NOTE: This means there their can be skips in the production patch numbers between release. E.g. v.0.0.9 might
be followed by v0.0.10-pre, v0.0.11-pre, v0.0.12-pre before a v0.0.12 appears as a production release.

Production and pre-releases should include Zip files of the compiled cli to be tested by `bash test_cmds.bash`.

Dev releases
------------

Dev release may happend from time to time as needed. They should always end in a '-dev' version number (e.g. v0.0.10-dev). 
They normally should not have any pre-compiled binaries to avoid confusion. They should be flagged as draft (pre-release)
on Github.

Making a release
----------------

1. Set the version number in PACKAGE.go (where PACKAGE is the name of the package, e.g. dataset is the name of the dataset
package so you'd change the version number in dataset.go).
2. Run `make clean`
3. Run `make test` and make sure they pass, if some fail document it if you plan to release it (e.g. GSheet integration not tested because...)
4. Run `make release`

You are now ready to go to Github and create a release. If you are uploading compiled versions upload the zip files in the _dist_
folder.



cmd/jsonmunge/jsonmunge.go
// jsonmunge is a command line tool that takes a JSON document and
// a Go text/template rendering the result. Useful for
// reshaping a JSON document or transforming into a new format,
// or filter for specific content.
//
// @author R. S. Doiel, <rsdoiel@caltech.edu>
package main

import (
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"os"
	"path"
	"text/template"

	// Caltech Library Packages
	"github.com/caltechlibrary/datatools"
	"github.com/caltechlibrary/tmplfn"
)

var (
	helpText = `%{app_name}(1) user manual | version {version} {release_hash}
% R. S. Doiel
% {release_date}

# NAME

{app_name} 

# SYNOPSIS

{app_name} [OPTIONS] TEMPLATE_FILENAME

# DESCRIPTION

{app_name} is a command line tool that takes a JSON document and
one or more Go templates rendering the results. Useful for
reshaping a JSON document, transforming into a new format,
or filter for specific content.

- TEMPLATE_FILENAME is the name of a Go text tempate file used to render
  the outbound JSON document

# OPTIONS

-help
: display help

-license
: display license

-version
: display version

-E, -expression
: use template expression as template

-i, -input
: input filename

-nl, -newline
: if true add a trailing newline

-o, -output
: output filename

-quiet
: suppress error messages


# EXAMPLES

If person.json contained

~~~
   {"name": "Doe, Jane", "email":"jd@example.org", "age": 42}
~~~

and the template, name.tmpl, contained

~~~
   {{- .name -}}
~~~

Getting just the name could be done with

~~~
    cat person.json | {app_name} name.tmpl
~~~

This would yield

~~~
    "Doe, Jane"
~~~

{app_name} {version}
`

	// Basic Options
	showHelp         bool
	showLicense      bool
	showVersion      bool
	showExamples     bool
	inputFName       string
	outputFName      string
	generateMarkdown bool
	generateManPage  bool
	quiet            bool
	newLine          bool

	// Application Specific Options
	templateExpr string
)


func main() {
	appName := path.Base(os.Args[0])
	version := datatools.Version
	license := datatools.LicenseText
	releaseDate := datatools.ReleaseDate
	releaseHash := datatools.ReleaseHash

	// Basic Options
	flag.BoolVar(&showHelp, "help", false, "display help")
	flag.BoolVar(&showLicense, "license", false, "display license")
	flag.BoolVar(&showVersion, "version", false, "display version")
	flag.BoolVar(&showExamples, "examples", false, "display example(s)")

	flag.StringVar(&inputFName, "i", "", "input filename")
	flag.StringVar(&inputFName, "input", "", "input filename")
	flag.StringVar(&outputFName, "o", "", "output filename")
	flag.StringVar(&outputFName, "output", "", "output filename")
	flag.BoolVar(&quiet, "quiet", false, "suppress error messages")
	flag.BoolVar(&newLine, "nl", false, "if true add a trailing newline")
	flag.BoolVar(&newLine, "newline", false, "if true add a trailing newline")

	// Application Specific Options
	flag.StringVar(&templateExpr, "E", "", "use template expression as template")
	flag.StringVar(&templateExpr, "expression", "", "use template expression as template")

	// Parse env and options
	flag.Parse()
	args := flag.Args()

	// Setup IO
	var err error

	in := os.Stdin
	out := os.Stdout
	eout := os.Stderr

	if inputFName != "" && inputFName != "-" {
		in, err = os.Open(inputFName)
		if err != nil {
			fmt.Fprintln(eout, err)
			os.Exit(1)
		}
		defer in.Close()
	}

	if outputFName != "" && outputFName != "-" {
		out, err = os.Create(outputFName)
		if err != nil {
			fmt.Fprintln(eout, err)
			os.Exit(1)
		}
		defer out.Close()
	}

	// Process Options
	if showHelp {
		fmt.Fprintf(out, "%s\n", datatools.FmtHelp(helpText, appName, version, releaseDate, releaseHash))
		os.Exit(0)
	}
	if showLicense {
		fmt.Fprintf(out, "%s\n", license)
		os.Exit(0)
	}
	if showVersion {
		fmt.Fprintf(out, "datatools, %s %s %s\n", appName, version, releaseHash)
		os.Exit(0)
	}

	if len(args) == 0 && templateExpr == "" {
		if err != nil {
			fmt.Fprintln(eout, err)
			os.Exit(1)
		}
	}

	var (
		tmpl *template.Template
	)
	if templateExpr != "" {
		// Read in and compile our templates expression
		tmpl, err = template.New("default").Funcs(tmplfn.AllFuncs()).Parse(templateExpr)
		if err != nil {
			fmt.Fprintln(eout, err)
			os.Exit(1)
		}
	} else {
		// Read in and compile our templates
		tmpl, err = template.New(path.Base(args[0])).Funcs(tmplfn.AllFuncs()).ParseFiles(args...)
		if err != nil {
			fmt.Fprintln(eout, err)
			os.Exit(1)
		}
	}

	// READ in the JSON document
	buf, err := ioutil.ReadAll(in)
	if err != nil {
		fmt.Fprintln(eout, err)
		os.Exit(1)
	}

	// JSON Decode our document
	var data interface{}
	decoder := json.NewDecoder(bytes.NewBuffer(buf))
	decoder.UseNumber()
	if err := decoder.Decode(&data); err != nil {
		fmt.Fprintln(eout, err)
		os.Exit(1)
	}

	// Execute template with data
	err = tmpl.Execute(out, data)
	if err != nil {
		fmt.Fprintln(eout, err)
		os.Exit(1)
	}
	if newLine {
		fmt.Fprintln(out, "")
	}
}
